### PR TITLE
feat(help): highlighted codeblocks

### LINKED ISSUE #20912: docs: syntax highlighting for (some) code examples
### Feature already in Vim?

No!

### Feature description

Now that we bundle a (mostly) working treesitter parser for help files, we can have nice things: properly syntax-highlighted code examples. This involves the following steps:

* [x] extend the vimdoc parser to support language annotation for code markers (e.g., `>lua`) and add the corresponding injections to the bundled queries (https://github.com/neovim/tree-sitter-vimdoc/pull/65)
* [x] add corresponding conceal rules to `syntax/help.vim` to fall back gracefully if treesitter is disabled (#21154)
* [x] adapt the HTML documentation generator script to turn annotated blocks into highlighted code blocks (using https://highlightjs.org? https://github.com/neovim/neovim.github.io/pull/303)
* [x] successively rewrite the (Neovim-specific) help pages to add the right language annotations (start with all files titled "NVIM REFERENCE MANUAL")
* [x] change `gen_vimdoc.py` to handle annotated `<pre>` codeblocks
* [x] add annotation to `<pre>` blocks in Lua and C docstrings
* [ ] profit!

### PR DESCRIPTION
**Pull request by clason**
Closes #20912 

* bump vimdoc parser and update queries with injections (may need further updating)
* change the syntax file to also recognize codeblock markers with language annotation (necessary)
* add annotations to static part of Nvim Reference Manual (see below)
* adapt `gen_help_html` to updated parser (with bonus syntax highlighting via `highlight.js`, needs https://github.com/neovim/neovim.github.io/pull/303)
* adapt `gen_vimdoc.py` to support language annotations (manually for `lua` and `vim` only)
* add annotations for Lua and C API source code (see below)
* add `news` entry

TODO: 

* backport `syntax/help.vim` change to `release-0.8`

<details><summary>source files with code examples</summary>

* [x] src/nvim/api/vim.c
* [x] src/nvim/api/win_config.c
* [x] src/nvim/api/extmark.c
* [x] src/nvim/api/command.c
* [x] src/nvim/api/autocmd.c
* [x] src/nvim/api/buffer.c
* [x] runtime/lua/vim/diagnostic.lua
* [x] runtime/lua/vim/ui.lua
* [x] runtime/lua/vim/lsp.lua
* [x] runtime/lua/vim/filetype.lua
* [x] runtime/lua/vim/fs.lua
* [x] runtime/lua/vim/keymap.lua
* [x] runtime/lua/vim/_editor.lua
* [x] runtime/lua/vim/shared.lua
* [x] runtime/lua/vim/lsp/handlers.lua
* [x] runtime/lua/vim/lsp/rpc.lua
* [x] runtime/lua/vim/lsp/buf.lua
* [x] runtime/lua/vim/lsp/codelens.lua
* [x] runtime/lua/vim/lsp/util.lua
* [x] runtime/lua/vim/lsp/diagnostic.lua
* [x] runtime/lua/vim/treesitter/query.lua
* [x] runtime/lua/vim/treesitter.lua

</details>

<details><summary>"Nvim reference manual" files</summary>

* [x] develop.txt
* [x] lsp-extension.txt
* [x] if_pyth.txt
* [x] channel.txt
* [x] job_control.txt
* [x] deprecated.txt
* [x] provider.txt
* [x] ui.txt
* [x] dev_style.txt
* [x] treesitter.txt
* [x] api.txt
* [x] support.txt
* [x] diagnostic.txt
* [x] lua.txt
* [x] luaref.txt
* [x] luvref.txt
* [x] news.txt
* [x] nvim.txt
* [x] remote_plugin.txt
* [x] nvim_terminal_emulator.txt
* [x] intro.txt
* [x] lsp.txt
* [x] term.txt
* [x] vim_diff.txt

</details>

### CODE REVIEW

**Comment by clason** [comment]
OK, this is looking reasonably good on my end so people interested can test this. @justinmk I'd appreciate it if you could give the HTML doc generation a good tire-kicking.

**Comment by lewis6991** [comment]
Looks good to me, however the language name is not concealed:

image

**Comment by lewis6991** [comment]
With treesitter (as it has language injection).

**Comment by clason** [comment]
Doesn't look like an injection to me ;)

**Comment by lewis6991** [comment]
image

**Comment by clason** [comment]
Do you have custom conceal?

**Comment by lewis6991** [comment]
Not to my knowledge. You mean a custom setting for `conceallevel`?

**Comment by clason** [comment]
Anything, really. I don't see how the `>` can be concealed while the language isn't.

**Comment by zeertzjq** [comment]
Won't this look ugly for user who don't have nvim-treesitter installed?

**Comment by lewis6991** [comment]
image

Being concealed. There must be something funky somewhere.

**Comment by clason** [comment]
> Won't this look ugly for user who don't have nvim-treesitter installed?

No, because I changed the syntax file to hide the language markers, too (see my second screenshot, which is with standard legacy highlighting).

**Comment by lewis6991** [comment]
Figured it out. I didn't have `syntax` enabled. It was using `conceal` from treesitter.

**Comment by clason** [comment]
Still, the queries in this PR should conceal the language as well -- which means you're not using the highlight queries from this PR but either the ones bundled on master or from nvim-treesitter.

**Comment by lewis6991** [comment]
Yep, I think it's pulling in `nvim-treesitter`.

**Comment by clason** [comment]
Can you update nvim-treesitter and try again?

**Comment by lewis6991** [comment]
All working now. Looks great!

**Comment by lewis6991** [comment]
For future extra credit, I think applying a different background for codeblocks would make it extra pretty.

**Comment by lewis6991** [comment]
Can't we just apply it to `codeblock>line`, and just make sure the injected highlights have blend?

**Comment by lewis6991** [comment]
With `:hi link @text.literl Folded` it sort of works:

image

**Comment by clason** [comment]
You'd probably want `@text.literal.help`, but that might be useful. In any case, this is something for (Neovim's future spiffy new default) colorscheme.

**Comment by clason** [comment]
Ok, this is basically done except for the TODOs listed as "to be discussed" @justinmk

**Comment by craigmac** [comment]
Off-topic but could we format each codeblock with Stylua or otherwise standardize the examples somehow? Maybe reuse the style config that Neovim core uses? I find that the examples tend to be all over the place and inconsistent use of APIs. For example, some examples have call parentheses and others don't. I think we should also replace usages of `nvim.api.X` in examples when a simpler/more convenient alternative comes out.

**Comment by clason** [comment]
This is indeed (very) off-topic here. PRs that improve our documentation are always welcome.

**Review by bfredl — APPROVED** [review]
nicely done!

There are a few instances of incorrect highlights (like vim-blocks that need to be changed from bare eval expressions to full vim script statements), but we can address these in follow up PR:s, as well fine-tuning default highlight etc etc.

File: runtime/doc/lua.txt [hunk 1]
```diff
@@ -360,16 +360,16 @@ treated specially.
 Vimscript v:lua interface                                         *v:lua-call*
 
 From Vimscript the special `v:lua` prefix can be used to call Lua functions
-which are global or accessible from global tables. The expression >
+which are global or accessible from global tables. The expression >vim
     v:lua.func(arg1, arg2)
```
Comment (bfredl) [inline runtime/doc/lua.txt:364]: this needs to be prefixed with `call`

File: runtime/doc/lua.txt [hunk 2]
```diff
@@ -360,16 +360,16 @@ treated specially.
...
-is equivalent to the Lua chunk >
+is equivalent to the Lua chunk >lua
     return func(...)
-where the args are converted to Lua values. The expression >
+where the args are converted to Lua values. The expression >vim
     v:lua.somemod.func(args)
```
Comment (bfredl) [inline runtime/doc/lua.txt:368]: same

File: runtime/doc/lua.txt [hunk 3]
```diff
@@ -360,16 +360,16 @@ treated specially.
...
-is equivalent to the Lua chunk >
+is equivalent to the Lua chunk >lua
     return somemod.func(...)
 
-In addition, functions of packages can be accessed like >
+In addition, functions of packages can be accessed like >vim
     v:lua.require'mypack'.func(arg1, arg2)
     v:lua.require'mypack.submod'.func(arg1, arg2)
```
Comment (bfredl) [inline runtime/doc/lua.txt:374]: samesis

File: runtime/doc/lua.txt [hunk 4]
```diff
@@ -378,11 +378,11 @@ is still valid as a function call of itself, in case require returns a useful
 value).
 
 The `v:lua` prefix may be used to call Lua functions as |method|s. For
-example: >
+example: >vim
     arg1->v:lua.somemod.func(arg2)
```
Comment (bfredl) [inline runtime/doc/lua.txt:382]: and this one

Comment (seandewar) [reply runtime/doc/lua.txt:382]: (`:eval` for this one)

**Review by lewis6991 — APPROVED** [review]

**Comment by lewis6991** [comment]
anything left blocking this being merged?

**Comment by clason** [comment]
> anything left blocking this being merged?

Feedback from @justinmk, and (possibly) merging the linked PR for neovim.io that adds `highlight.js` support for user docs.

**Pull request merged**